### PR TITLE
Feat/add num gate types

### LINKED ISSUE #40: Add a function which determines the number of gates of each type in a quantum circuit

### PR DESCRIPTION
**Pull request by npoirie1**
Adds a function that determines the number of gates of each type in a circuit. This pull request also adds functions that compute the total number of gates in a circuit and the circuit depth.

Fixes #40

### CODE REVIEW

File: src/core/quantum_circuit.jl [hunk 1]
```diff
@@ -505,3 +505,139 @@ function simulate_shots(c::QuantumCircuit, shots_count::Int = 100)
...
+
+
+julia> get_gate_counts(c)
+Dict{String, Int64} with 2 entries:
+  "h"  => 3
+  "cx" => 1
+
+```
+"""
+function get_gate_counts(circuit::QuantumCircuit)
```
Comment (plerouxanyon) [inline src/core/quantum_circuit.jl:544]: Please provide a return-type annotation. I think `AbstractDict{<:AbstractString, <:Integer}` would match the intent of the function.

File: src/core/quantum_circuit.jl [hunk 2]
```diff
@@ -505,3 +505,139 @@ function simulate_shots(c::QuantumCircuit, shots_count::Int = 100)
...
+               
+
+
+
+julia> get_num_gates(c)
+3
+
+```
+"""
+function get_num_gates(circuit::QuantumCircuit)
```
Comment (plerouxanyon) [inline src/core/quantum_circuit.jl:586]: Please provide a return-type annotation. I think `Integer` would match the intent of the function.

File: src/core/quantum_circuit.jl [hunk 3]
```diff
@@ -505,3 +505,139 @@ function simulate_shots(c::QuantumCircuit, shots_count::Int = 100)
...
+          
+
+
+
+julia> get_depth(c2)
+1
+
+```
+"""
+function get_depth(circuit::QuantumCircuit)
```
Comment (plerouxanyon) [inline src/core/quantum_circuit.jl:641]: Also, what is the intended use case for `get_depth`? The hardware might not use an equal number of clock cycles for all operations, and the "hardware"-depth might be different than the logical depth. If the intent is to get the logical depth, then perhaps rename the function to `get_logical_depth`.

**Comment by plerouxanyon** [comment]
@npoirie1  I also see that there are merge conflicts with the main branch

If all the comments are addressed, it will be good to merge.

File: src/core/quantum_circuit.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (npoirie1) [reply src/core/quantum_circuit.jl]: A return-type annotation has been added.

File: src/core/quantum_circuit.jl (hunk 2 above)
Comment (npoirie1) [reply src/core/quantum_circuit.jl]: A return-type annotation has been added.

File: test/circuit.jl [hunk 4]
```diff
@@ -78,3 +78,18 @@ end
     c = QuantumCircuit(qubit_count = 2, bit_count = 0)
     @test_throws DomainError push_gate!(c, control_x(1, 3))
 end
+
+@testset "get_gate_counts" begin
```
[the remark this answers is not in the document]
Comment (npoirie1) [reply test/circuit.jl:95]: A separate testset was created for each function.

File: src/core/quantum_circuit.jl (hunk 3 above)
Comment (npoirie1) [reply src/core/quantum_circuit.jl]: The function `get_depth` was renamed to `get_logical_depth`.

**Comment by npoirie1** [comment]
@plerouxanyon, all the comments have been addressed and the merge conflicts have been resolved.

**Pull request merged**